File: event/2024/q17.jl
#!/usr/bin/env julia
using Graphs

const Pos = CartesianIndex{2}


function distance(a::Pos, b::Pos)
    abs.((b - a).I) |> sum
end


function parsenotes(text)
    grid = stack(split(text), dims=1)
    stars = findall(==('*'), grid)
    return (grid, stars)
end


function part1(data)
    grid, stars = parsenotes(data)
    n = length(stars)
    w = [distance(stars[i], stars[j]) for i in 1:n, j in 1:n]
    g = complete_graph(n)
    n + boruvka_mst(g, w).weight
end


part2 = part1


function part3(data, W=5)
    grid, stars = parsenotes(data)
    n = length(stars)
    w = [distance(stars[i], stars[j]) for i in 1:n, j in 1:n]
    w[w .> W] .= 0
    g = SimpleGraph(w)
    qs = connected_components(g)
    res = [length(q) + boruvka_mst(g[q], w[q,q]).weight
        for q in qs]
    sort(res, rev=true)[1:3] |> prod
end


data = raw"
*...*
..*..
.....
.....
*.*..
"
@assert part1(data) == 16


data = raw"
.......................................
..*.......*...*.....*...*......**.**...
....*.................*.......*..*..*..
..*.........*.......*...*.....*.....*..
......................*........*...*...
..*.*.....*...*.....*...*........*.....
.......................................
"
@assert part3(data) == 15624


data = readchomp("q17_p1.txt")
println("part1: ", part1(data))
data = readchomp("q17_p2.txt")
println("part2: ", part2(data))
data = readchomp("q17_p3.txt")
println("part3: ", part3(data))

File: event/2024/q17.py
#!/usr/bin/env python
from collections import deque


def part1(data):
    grid = {(y,x):c for y,s in enumerate(data.strip().splitlines())
        for x,c in enumerate(s)}
    stars = [p for p,c in grid.items() if c == '*']

    wei = [[0] * len(stars) for _ in range(len(stars))]
    for i,(y,x) in enumerate(stars):
        for j,(v,u) in enumerate(stars[i+1:], i+1):
            wei[i][j] = abs(y-v) + abs(x-u)
            wei[j][i] = wei[i][j]

    ans = 1
    seen = {0}
    while len(seen) < len(stars):
        p = sorted((w, j) for i in seen for j, w in enumerate(wei[i])
            if i != j and j not in seen)
        w,j = p[0]
        ans += w + 1
        seen.add(j)
    return ans


def part2(data):
    return part1(data)


def part3(data):
    grid = {(y,x):c for y,s in enumerate(data.strip().splitlines())
        for x,c in enumerate(s)}
    stars = [p for p,c in grid.items() if c == '*']

    wei = [[0] * len(stars) for _ in range(len(stars))]
    for i,(y,x) in enumerate(stars):
        for j,(v,u) in enumerate(stars[i+1:], i+1):
            wei[i][j] = abs(y-v) + abs(x-u)
            wei[j][i] = wei[i][j]

    cons = list()
    avail = set(range(len(stars)))
    done = set()
    while avail:
        s = avail.pop()
        seen = {s}
        res = 1
        while avail:
            p = sorted((w, j) for i in seen for j, w in enumerate(wei[i])
                if i != j and w < 6 and j not in seen and j not in done)
            if not p: break
            w,j = p[0]
            res += w + 1
            seen.add(j)
        cons.append(res)
        done |= seen
    a,b,c = sorted(cons)[-3:]
    ans = a * b * c
    return ans



data = """
*...*
..*..
.....
.....
*.*..
"""
assert part1(data) == 16
assert part2(data) == 16

data = """
.......................................
..*.......*...*.....*...*......**.**...
....*.................*.......*..*..*..
..*.........*.......*...*.....*.....*..
......................*........*...*...
..*.*.....*...*.....*...*........*.....
.......................................
"""
assert part3(data) == 15624


data = open('q17_p1.txt').read()
ans = part1(data)
print('part1:', ans)

data = open('q17_p2.txt').read()
ans = part2(data)
print('part2:', ans)

data = open('q17_p3.txt').read()
ans = part3(data)
print('part3:', ans)
